src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -471,19 +489,114 @@ void enforceMemberNotNull(SyntaxNode? syntaxOpt, LocalState state)
                 }
             }
 
+            void checkMemberStateOnConstructorExit(MethodSymbol constructor, Symbol member, LocalState state, int thisSlot, Location? exitLocation)
+            {
+                var isStatic = !constructor.RequiresInstanceReceiver();
+                if (member.IsStatic != isStatic)
+                {
+                    return;
+                }
+
+                // This is not required for correctness, but in the case where the member has
+                // an initializer, we know we've assigned to the member and
+                // have given any applicable warnings about a bad value going in.
+                // Therefore we skip this check when the member has an initializer to reduce noise.
+                if (HasInitializer(member))
+                {
+                    return;
+                }
+                TypeWithAnnotations fieldType;
+                FieldSymbol? field;
+                switch (member)
+                {
+                    case FieldSymbol f:
+                        fieldType = f.TypeWithAnnotations;
+                        field = f;
+                        break;
+                    case EventSymbol e:
+                        fieldType = e.TypeWithAnnotations;
+                        field = e.AssociatedField;
+                        if (field is null)
+                        {
+                            return;
+                        }
+                        break;
+                    default:
+                        return;
+                }
+                if (field.IsConst)
+                {
+                    return;
+                }
+                if (fieldType.Type.IsValueType || fieldType.Type.IsErrorType())
+                {
+                    return;
+                }
+                if (!fieldType.NullableAnnotation.IsNotAnnotated())
+                {
+                    return;
+                }
+                var symbol = member switch
+                {
+                    FieldSymbol { AssociatedSymbol: PropertySymbol p } => p,
+                    _ => member
+                };
+                var annotations = symbol.GetFlowAnalysisAnnotations();
+                if ((annotations & FlowAnalysisAnnotations.MaybeNull) != 0
+                    || (annotations & FlowAnalysisAnnotations.AllowNull) != 0)
+                {
+                    // We assume that if a member allows null input then the user
+                    // does not care that we exit at a point where the member might be null.
+                    return;
+                }
+                var slot = GetOrCreateSlot(symbol, thisSlot);
+                if (slot < 0)
+                {
+                    return;
+                }
+
+                var memberState = state[slot];
+                var badState = fieldType.Type.IsPossiblyNullableReferenceTypeTypeParameter() ? NullableFlowState.MaybeDefault : NullableFlowState.MaybeNull;
```

Comment:
This seems like it could be stronger if we checked for definitely null like `IsNullableTypeOrTypeParameter` and used `MaybeNull` for that and `MaybeDefault` for the rest.

Reply:
Hmm, seems like this pattern exists a bit in the code so probably okay.